Add "Show All Diagnostics"

davisp opened the pull request:
Closes #4901

Release Notes:

- Added ability to show all diagnostics rendered in an editor

Some quick examples of what these look like when rendered:

<img width="1240" alt="Screenshot 2024-09-02 at 10 09 21 AM" src="https://github.com/user-attachments/assets/f5cbd3bd-0a7a-4fc4-aa88-9fbbb7da1182">

<img width="1240" alt="Screenshot 2024-09-02 at 10 09 48 AM" src="https://github.com/user-attachments/assets/8a24b4af-707a-4627-bd8c-9c3a6c01c21b">

This PR changes Editor::active_diagnostics from `Option<ActiveDiagnosticsGroup>` to `HashMap<usize, ActiveDiagnosticsGroup>` so that we can track multiple diagnostics groups. Toggling all diagnostics on or off displays or hides all active diagnostics in the current editor. GoToDiagnostic and GoToPrevDiagnostic still support moving through each diagnostic individually while also maintaining the existing behavior single diagnostic view when Show All Diagnostics is off.

There aren't any tests associated with this PR because I frankly have no idea where or how to add tests for GUI application. Given that `cargo test --all-targets` only runs 22 tests, I'm guessing I'm poking the test button wrong. In any case, manual testing and normal usage appear to show that its working as intended.

Some notes for Zed devs:

* I preemptively added the ability to style background colors for diagnostic messages based on severity. I have no idea if there's a process for adding theme-able elements, and the default colors I used kinda clash with my current theme. If there's a process for that I've got no problem reverting those changes.
* I picked `F9` as the hotkey for the toggle. I didn't see it bound anywhere else. I have no idea if there's a better default based on other editors or not.
* Related to the keybinding, I also added a menu item for the toggle, but it doesn't change state and is awkwardly greyed out when not focused on an editor. I'm guessing that setting probably needs to be moved out of the Editor instance and into some global configuration thinger.
* Updating diagnostic displays on every keypress can get a bit wonky. I probably need to add a delayed/debounce task thing that waits for some configurable amount of time after the last refresh request to avoid the visual noise.

davisp committed:
Refactor Editor::active_diagonstics

This changes the active_diagnostics field from an Option to a HashMap
from group_id to ActiveDiagnosticGroup and adds a focused_diagnostic to
track the currently selected group for reference by
editor::GoToDiagnostic.

The next step is to create a field that toggles whether all diagnostics
should be shown or hidden while editor::GoToDiagnostic still cycles
focus through each diagnostic.

davisp committed:
First pass at hard coding show all diagnostics

The boolean is current hard coded to on while I check to see if the
basics of showing all diagnostics seem to work.

davisp committed:
Allow for styling diagnostic messages

davisp committed:
Add action to toggle showing all diagnostics

This defaults to using F9 to toggle showing all diagnostics. I have no
idea if VSCode has a button for this. I couldn't find anything obvious
on Google.

davisp committed:
Return default to not showing all diagnostics

Now that the toggle works we can revert the default to off.

bitfield commented on crates/theme/src/schema.rs:
```diff
@@ -401,6 +401,25 @@ pub struct ThemeColorsContent {
+    ///
+    /// These colors affect the background of diagnostics that are rendered
+    /// inline with code.
+    #[serde(rename = "editor.diagbostic_error_background")]
```
```suggestion
    #[serde(rename = "editor.diagnostic_error_background")]
```

etc

davisp replied:
These are fixed now.

davisp committed:
Fix typos in theme setting names

davisp commented:
@bitfield Thanks for the review! And spot on with this being not quite right but hopefully close enough to the ballpark that a core dev can point me in the right direction to get it across the finish line. I'll try and find some time this weekend to work on getting the configuration settled along with a debounce configuration added as I think those are likely the two biggest issues at this point.

I can't decide which direction to go with the theming stuff. I totally agree with you about getting it in will likely be easier without it, but its not much code so I think I'll just leave it for now. I'd be more than happy to remove it if asked, but if it all goes in at once that seems slightly preferable.

davisp committed:
Add configuration key show_diagnostics_inline

This makes whether diagnostics are automaticall shown inline take a
configurable default. For now I've left it turned off. This way users
don't have to remember to toggle the value each time an editor is
created.

beard-programmer commented:
This feature is a dealbreaker.
Sticking with Goland for now until its done - which hopefully will happen asap bc goland sucks.

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -10088,17 +10120,40 @@ impl Editor {
+                return;
+            }
+
+            let (primary_range, primary_message) = primary.unwrap();
```
> .unwrap()

Please rewrite this as `match primary` or any other way that does not introduce another potential panic.
At worst, an `.expect("message")` has to be used instead of `.unwrap()` but it's not the case here.

davisp replied:
Generally curious here to further my Rust knowledge. Looking at this code I certainly see ways to write it better today, but given that is_none/return a few lines above, this unwrap is obviously not going to panic.

The obvious counter argument is "what's true today isn't necessary necessarily true tomorrow" after a separate PR changes the logic and leaves a hole where the unwrap could trigger. Which, fair point! Especially on a project as large as Zed.

I just haven't (yet?) seen a "Thou shalt not unwrap!" commandment adopted anywhere which I'm starting to entertain as a possibility for a core library I've been working on. Speaking of, does Zed have a lint for this?

SomeoneToIgnore commented on crates/editor/src/editor.rs:
```diff
@@ -10119,24 +10179,34 @@ impl Editor {
+
+    fn unfocus_diagnostics(&mut self, cx: &mut ViewContext<Self>) {
+        self.focused_diagnostic = None;
+        self.refresh_active_diagnostics(cx);
```
I find it quite suspicious that we have to do something relatively complex (as iterating over all diagnostics and updating them) when they go in/out of focus.

Curious, why do we need that? 
Could we avoid any diagnostics update and use the old data instead?

davisp replied:
Whoops. That's totally me making another "See, I'm totally not a UI dev" mistake. The focus/unfocus is a bad name I picked for changing which of the diagnostics `F8` is currently pointing at. I tried to keep it so that if you were searching around F8 style and then toggled the F9 version on and off, the F8 progress/state wouldn't be lost. Oh and also so that F8 can jump to next diagnostic even when F9 is on, which I think is where focus came from (i.e., scroll next one into view). In hindsight, I could see how a real UI person would immediately think window focus/unfocus instead.

As to the specific question on optimizations, my vague memory is "I could optimize this by duplicating some of the logic in refresh_active_diagnostics, or I could just take the (possible, unmeasured, maybe zero) performance hit in completely re-running the refresh logic." Also, this unfocus is only ever called when users dismiss the active F8 style diagnostic and we're already running refresh_active_diagnostics on every key stroke, so :shrug: on that performance hit.

SomeoneToIgnore commented on crates/editor/src/editor_settings.rs:
```diff
@@ -261,6 +262,16 @@ pub struct EditorSettingsContent {
+    /// `show_diagnostics_inline` but I don't know if that's kosher.
+    ///
+    /// Default: false
+    pub show_diagnostics_inline: Option<bool>,
```
Actually, something more broken seems to be in place here: when I press and hold F9, I would expect the diagnostics to toggle, but this does not happen.

Also, seems that not only the toggling off is broken: if I open the buffer and keep on pressing F9, not much of an update happens.
This needs a fix and a test.

SomeoneToIgnore commented on crates/zed/src/zed/app_menus.rs:
```diff
@@ -128,6 +128,7 @@ pub fn app_menus() -> Vec<Menu> {
                 MenuItem::action("Terminal Panel", terminal_panel::ToggleFocus),
                 MenuItem::separator(),
                 MenuItem::action("Diagnostics", diagnostics::Deploy),
+                MenuItem::action("Toggle Show All Diagnostics", editor::actions::ToggleDiagnostics),
```
I think this action needs a rename: it's not clear that we toggle all diagnostics to be shown inline.
What is "toggling a diagnostic" even mean?

`ToggleShowAllDiagnostics` is better than the original, but can we do even better?
The new name also has to match the settings name.

SomeoneToIgnore reviewed (commented):
Sorry for taking quite a time before submitting the review.
Hopefully, it can go faster now after I've got all the context processed.

> There aren't any tests associated with this PR because I frankly have no idea where or how to add tests for GUI application. 

I think we're fine with not having any UI tests, but we might create one in the `editor_test` and check that on toggling the diagnostics back and forth, the state of the editor changes (new blocks appear).
Right now it's broken, as I cannot toggle off these diagnostics by just pressing F9, I have to also press another button like `down` for them to disappear — see the video below.

> I'm guessing I'm poking the test button wrong

Yes, for starters, there's an `--all`/`--workspace` flag to run all tests in all modules.
We also use `nextest` to run the tests, in case you want to replicate CI entirely (not mandatory): https://github.com/zed-industries/zed/blob/2d6ac6fec334a5ed2763d3422f915698e948cb69/.github/actions/run_tests/action.yml#L23 

> I preemptively added the ability to style background colors for diagnostic messages based on severity. I have no idea if there's a process for adding theme-able elements, and the default colors I used kinda clash with my current theme. If there's a process for that I've got no problem reverting those changes.

Yes, let's revert that for simplicity for now and concentrate on making things less ugly first. The designers will polish it later.

> I picked F9 as the hotkey for the toggle. I didn't see it bound anywhere else. I have no idea if there's a better default based on other editors or not.

Seems fine

> Related to the keybinding, I also added a menu item for the toggle, but it doesn't change state and is awkwardly greyed out when not focused on an editor. I'm guessing that setting probably needs to be moved out of the Editor instance and into some global configuration thinger.

I think it's fine? It's about showing the diagnostics the editor, after all, what should we do when we have an assistant panel focused and calling this action?
If you really want to explore that path, there's always a `workspace::` namespace for actions that are global.

> Updating diagnostic displays on every keypress can get a bit wonky. I probably need to add a delayed/debounce task thing that waits for some configurable amount of time after the last refresh request to avoid the visual noise.

Undeniably! It is quite flickering right now as you might see in my video below.

----------------------------------------

My main issue with the PR is how large amounts of diagnostics are displayed.
Consider a few examples:

* offsets are messed up
![offsets](https://github.com/user-attachments/assets/6d6d2d54-fb65-4932-bdf5-3171c55d21bc)
and it does not get better:
![worse_offsets](https://github.com/user-attachments/assets/d120c2e2-07cb-4d15-b31d-145e774f6c9d)

While this is a Rust code, I also strongly suggest to check out JS/TS errors that might be quite abundant and multiline. 
Diagnostics text lines can also be very long, what will current impl do then? Scroll, wrap? 
We need some verification there.

![offsets_again](https://github.com/user-attachments/assets/532756c2-f3f7-47ce-aa4f-2ef304fde179)

* F9 does not work reliably: see my comments below, it needs a test.
This leaves the feeling that updates are quite slow, since nothing is refreshed onscreen.

* I found it confusing that `ESC` does not work and does not toggle off these diagnostics — most (if not all) of the editor blocks in Zed (e.g. diff views and inline assistant code blocks) disappear after it's pressed and I think we should do it for the new blocks too.

* As noted in the PR message, flickers are quite hazardous and we need to fix them:

https://github.com/user-attachments/assets/5ee76b88-104f-4a6b-9e8a-cc5959c8f6c5

* You may also notice horrendous offsets and the fact that certain messages are identical:
![image](https://github.com/user-attachments/assets/ba4574dc-908f-4253-9631-095a4f37f392)

It's not at all clear which of `Syntax Error: expected an item` is related to where.
Similarly, it's not clear at all which diagnostics group belongs to where: right now, there's a logic, making "go to next/previous diagnostic" (F8/shift-F8) to jump between primary diagnostics only — when doing that, Zed hides non-primary diagnostics from the old group and shows the new ones:

https://github.com/user-attachments/assets/a7e176b0-477e-422f-94cd-55089b546d31

I like this a lot, as it helps to group related diagnostics and show only the relevant ones.
In our case, where we have to show everything, it gets quite hard to understand why certain diagnostics are not navigated to with F8 and which group is currently processed.
It feels like we have to highlight the entire group when focusing a particular primary diagnostic.

-----------------

So, overall, besides things being badly named, actions not always working and text flickering, there seems to be a large task of understanding how to group the diagnostics and display many of them at once without visual confusion.

Back in the days, I've tried to do something similar as this PR does, but inside a standalone editor.
I've added a toggle button and displayed all multiline and same-place-located diagnostics folded: https://github.com/zed-industries/zed/pull/14515/files#diff-a3da3181e4ab4f73aa1697d7b6dc0caa0c17b2a187fb83b076dfc0234ec91f54R12715

I do not think it's good enough, but much better than the current chaos — I think it's the core (and only) design thing we have to resolve during this PR before merging.

bitfield commented:
> Diagnostics text lines can also be very long, what will current impl do then? Scroll, wrap?

One thing that's very neat about Error Lens is that it shows the diagnostic _on the same line_ (at the end), not between lines:

![](https://i.sstatic.net/nGrYS.png)

This is important, because it means the code you're looking at does not move, and it doesn't jump around as the diagnostics change. I think we'd need to do the same thing, otherwise this feature would be quite visually jarring to use. This also solves the offset issue.

The downside is that maybe you can't see the full text of the diagnostic because it doesn't wrap. But the trade-off is worth it. After all, if you want the _full_ diagnostic you can always hover. The key speedup from Error Lens is that you can usually see enough of the error message to be able to fix it.

jansol commented:
> One thing that's very neat about Error Lens is that it shows the diagnostic on the same line (at the end), not between lines

The arguments for having the first line of the diagnostic at the end of the line are good, but I for one quite like having them as a new block (old Xcode-style) so I would appreciate at least having that as an option.

SomeoneToIgnore commented:
This starts to get stale so I'll close it and emphasize again that the hardest part here is not even technical: we need to understand how multiple long line diagnostics at the same line are displayed; same as other pathological cases.

The PR looks quite good overall, so it might serve as a base for future improvements, if anyone wants to take a stab later.

davisp commented:
Apologies for taking so long to respond, I ended up getting busy with work and haven't gotten around to updating this PR as far as I would have liked.

For some top line context, I've been using this branch of zed as my daily driver for editing Rust full time and I am 100% on board with the fact that this is absolutely not ready for prime time. The diagnostic display is both atrocious and bumpy. Not being a UI dev at all I hadn't even considered the flickering issue in regards to accessibility stuff which is enough to ixnay this PR on its own IMO.

Also, given that context, I wouldn't be using Zed at all without this branch. I think there are enough other folks echoing that sentiment that I don't really need to underline how important *something* of this nature is desired whether or not there's agreement on the specifics (inline vs blocks).

And now on to responding to the review to provide some more context for others.

> My main issue with the PR is how large amounts of diagnostics are displayed.

I'm 100% in agreement here. However, I want to be clear that the current state of affairs weird duplicates and their sometimes odd positioning is not due to this PR. (Well, other than I didn't change anything to fix the issue I guess).

This entire PR can be succinctly summarized as "Rather than having to press F8 to cycle through diagnostics individually, let me toggle them all on at once". That is to say, in the examples shown, those are all real diagnostics that would be cycled through and displayed exactly the same. I've wondered a couple times while using this if I didn't introduce a duplications somewhere when its egregious, but each time I turn off the show all, I still cycle through all of them.

That's to say, those are the diagnostics as they exist for the F8 cycling behavior. Any work on de-duplicating those will likely require some spelunking between the Editor and however the plumbing to LSP land works.

One other thing I'll point out is that the reason that this branch shows diagnostics as blocks is simply because that's how F8 works. I have absolutely no idea how one would do the magical "show diagnostic text on the same line" similar to inlay hints or whatever. I assume I could figure it out eventually, but in terms of time my crap approach was enough to get me using Zed full time so that was good enough for me.

> * F9 does not work reliably: see my comments below, it needs a test.
>   This leaves the feeling that updates are quite slow, since nothing is refreshed onscreen.

I think this is two issues after having used this quite a lot. F9 itself works fine. I've never had it not work. However, I have had a few encounters where the diagnostics got so bad I had to toggle Show All to fix whatever I was working on before re-enabling due to latency and other issues. Performance is absolutely part of why I don't think this PR is any where near mergeable.

> * I found it confusing that `ESC` does not work and does not toggle off these diagnostics — most (if not all) of the editor blocks in Zed (e.g. diff views and inline assistant code blocks) disappear after it's pressed and I think we should do it for the new blocks too.

I believe this fundamentally misunderstands what most folks want out of this feature. The entire point is that these diagnostics are shown all the time forever and for always (well, except for when turning off the feature). But they are absolutely not a modal to be dismissed. Think of it like an instantaneous compiler loop. If the compiler says something won't compile, it gets highlighted in bright colors with hopefully enough context presented to address the issue without leaving the editor (or slightly less worse, having to hover over a target). Whether that's inline a la Code Lens style, or as blocks is secondary (and should probably be configurable if I'm being honest).

> As noted in the PR message, flickers are quite hazardous and we need to fix them:

Point absolutely acknowledged! I'm not a UI dev so I didn't even consider the UI/Accessibility aspect there, but that's yet another nail in this PR's coffin.

I'll respond to a few of the other inline comments as well, though given this specific PR is dead in the water, I'll skip the obvious changes as there's nothing I disagree with in the requested changes.

davisp commented on crates/theme/src/default_colors.rs:
```diff
@@ -78,6 +78,11 @@ impl ThemeColors {
             editor_active_wrap_guide: neutral().light_alpha().step_8(),
             editor_indent_guide: neutral().light_alpha().step_5(),
             editor_indent_guide_active: neutral().light_alpha().step_6(),
+            editor_diagnostic_error_background: neutral().light_alpha().step_3(),
```
Ah, a good note for any future PRs. I'm absolutely on board leaving this to the professionals. This was nothing more than "Well, I see how it works so I can add it" sort of code.

davisp commented on crates/editor/src/editor.rs:
```diff
@@ -10038,45 +10048,67 @@ impl Editor {
-        if let Some(active_diagnostics) = self.active_diagnostics.as_mut() {
-            let buffer = self.buffer.read(cx).snapshot(cx);
-            let primary_range_start = active_diagnostics.primary_range.start.to_offset(&buffer);
-            let is_valid = buffer
```
Right, so this just got simplified out due to how I added support for showing all diagnostics. The basic idea here is a standard set update. I have an old set of diagnostics (whatever is currently being shown) and the new set (whatever the LSP has just sent us). The idea is only hide diagnostics that no longer exist while only adding newly encountered diagnostics. The standard optimization over "hide all old ones, show all new ones" sort of optimization.

The `is_valid` is just doing the same thing except since it only ever cared about a single diagnostic group at once its doing the "what to remove/what to add" logic directly at the individual diagnostic level rather than at the group_id level that I added for showing all diagnostics. I just reused the same logic since it simplified things a bit at the cost of storing that list in the `self.active_diagnostics` HashMap.

SomeoneToIgnore commented:
No worries and thank you for testing that yourself, that is the main part (followed by the ability to fix the issues, of course).
I was not so swift to provide the feedback anyway.

Regarding the F8 feedback and diagnostics displayed, I would wait until https://github.com/zed-industries/zed/tree/overhaul-multibuffer lands to see how the new multi buffer will behave: if we can extend its excepts easier and merge them on the fly, we might be able to show all diagnostics at once, jump between them and be quite close to regular buffer(s).
That is not the replacement, but somehow an alternative to some people, at least.

Roger the ESC idea, albeit I suspect the future user feedback will converge us into that mode, just later.
I'd keep that in mind when renaming the settings, and added the enum "always"/"never" instead of a boolean as it is now.
So that later someone could fit that behavior into the setting easier.

I think if we fix the

> The diagnostic display is both atrocious and bumpy. Not being a UI dev at all I hadn't even considered the flickering issue in regards to accessibility stuff which is enough to ixnay this PR on its own IMO.

and whatever brings "stuck" F9 feeling, this will be a perfect first iteration on the topic, given that we also add some logical "unit" tests that show/hide the diagnostics and check the editor (?) data changes (so no UI things).

I would also not put much work into customizing the diagnostics view, but concentrated on grouping and layouting them so that it's normal to read the code somehow.
If we group the items "logically", that will cut off half of the sinister appeal at least.
